DimensionPreloadEvent

goreacraft opened the pull request:
This is a remake of the commit i made earlier : #3758 and following mezz and diesieben07 suggestions

In my case this event is important for server management purpose and will allow the creation of customized dimension for mining , market, events for example and allow full customization on how this dimensions are generated and stored (with mods server side only). 

To do this one can register its DimensionType with the custom WorldProvider in server init and then detect when this dimension gets instantiated and push the desired WorldServer object instead of allowing Forge to cast it to WorldServerMulti.

goreacraft committed:
This is a remake of the commit i made earlier : #3758 and following mezz and diesieben07 sugestions

goreacraft committed:
This is a remake of the commit i made earlier : #3758 and following mezz and diesieben07 sugestions

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

# Conflicts:
#	src/main/java/net/minecraftforge/event/world/DimensionPreloadEvent.java

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

# Conflicts:
#	src/main/java/net/minecraftforge/event/world/DimensionPreloadEvent.java

liach commented on src/main/java/net/minecraftforge/event/world/DimensionPreloadEvent.java:
```diff
@@ -0,0 +1,61 @@
+ * This event can be used to override instantiated dimensions <br>
+ * by using {@link #setAlternativeDimension }
+ */
+public class DimensionPreloadEvent extends Event {
```
braces on the new line

goreacraft replied:
done.. sorry for that

liach reviewed (changes requested):
All the braces need to get fixed.

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

mezz commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -216,20 +216,24 @@ public static void initDimension(int dim)
-        //WorldSettings worldSettings = new WorldSettings(overworld.getWorldInfo());
 
-        WorldServer world = (dim == 0 ? overworld : (WorldServer)(new WorldServerMulti(mcServer, savehandler, dim, overworld, mcServer.theProfiler).init()));
+        WorldServer world = (dim == 0 ? overworld : ( alterantiveDimension != null ? alterantiveDimension : ((WorldServer)(new WorldServerMulti(mcServer, overworld.getSaveHandler(), dim, overworld, mcServer.theProfiler).init()))));
```
This line has two ternary statements in it now, can you break it up into if / else if / else?

goreacraft replied:
indeed... the readability is much better with clasic if statements

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

mezz commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -216,20 +216,34 @@ public static void initDimension(int dim)
+            }
+            else
+            {
+                world = (WorldServer) (new WorldServerMulti(mcServer, overworld.getSaveHandler(), dim, overworld, mcServer.theProfiler).init());
```
Can you make this section this way?
```java
WorldServer world;
if (dim == 0)
{
    world = overworld;
}
else if (alternativeDimension != null)
{
    world = alterantiveDimension;
}
else
{
    world = new WorldServerMulti(mcServer, overworld.getSaveHandler(), dim, overworld, mcServer.theProfiler);
    world.init();
}
```

goreacraft replied:
sure.. and you just showed me that i misspelled "alterantiveDimension" too :d

goreacraft committed:
Added Cancelable DimensionPreloadEvent. By canceling this event will prevent the dimension from being initialized and then you can use.
DimensionManager.setWorld to add your own custom dimension instead. This will make possible to create a proper multiverse server side plugin

This is a remake of the commit i made earlier : #3755

goreacraft committed:
Added Cancelable DimensionPreloadEvent. By canceling this event will prevent the dimension from being initialized and then you can use.
DimensionManager.setWorld to add your own custom dimension instead. This will make possible to create a proper multiverse server side plugin

This is a remake of the commit i made earlier : #3755

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

# Conflicts:
#	src/main/java/net/minecraftforge/common/DimensionManager.java

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

# Conflicts:
#	src/main/java/net/minecraftforge/common/DimensionManager.java

goreacraft commented:
I have made a test mod with some basic functionality, commands to test:
* _/mining_ (when run first time will create a new dimension) and then this command will teleport to mining dimension 
* _/mining info_ - gives information about dimension
* _/mining [int]_ - adding an int argument to the command will work like a basic tp command between dimensions so '/mining 0' will tp back to overworld.
* _/mining reset_ - will make it reset the registered mining id to make it possible to generate new one (will not delete the folder)

It is buggy but should be fine for testing purposes i think. If it is necessary to improve it let me know (i need to create a separate class file to make the custom provider work)

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

# Conflicts:
#	src/test/java/net/minecraftforge/test/MiningDimensionTest.java

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

# Conflicts:
#	src/test/java/net/minecraftforge/test/MiningDimensionTest.java

mezz commented on src/test/java/net/minecraftforge/test/MiningDimensionTest.java:
```diff
@@ -0,0 +1,640 @@
+    public void onServerStarting(FMLServerStartingEvent event)
+    {
+        event.registerServerCommand(new MiningDimensionTest.DimensionTestCommand());
+        System.out.println("MiningDimensionTest Registered  command /mining <int>");
```
Please use the logger you get from `FMLPreInitializationEvent#getModLog` instead of `System.out`

goreacraft replied:
Sorry for delay, i will update this asp and improve it a little bit

goreacraft replied:
for now i will push only the changes you requested, i will leave the improvements for later as don't have the time at this moment

mezz commented:
The commands looks useful, but what action do I take to test your new event?

goreacraft committed:
Merge remote-tracking branch 'origin/1.11.x' into 1.11.x

# Conflicts:
#	src/test/java/net/minecraftforge/test/MiningDimensionTest.java

goreacraft commented:
To actually take full advantage of this event i will need to upload an entire multiverse mod so instead i just made small test that will create a customized mining dimension (custom name, custom save location, keepinventory, allday, worldtype amplified, basically a bunch of random options just for testing) , so to actually test  this is to see if the changes applied correctly and there is a beautiful mining dimension generated when using the command /mining

goreacraft commented:
Hello, was wondering if there is a reason why this has not been accepted yet, do i need to make any changes or add more explanations?

diesieben07 commented:
- The test mod is way too big.
- I still have not seen why this is needed. Your usecase is a "mining dimension", what is that supposed to be? Why do you need a custom World implementation for that?

goreacraft commented:
The "mining dimension" is just an example i chosen randomly and there are way more situations where a server can benefit from the flexibility added by this event. 
Maybe i am missing something but do you know of another way to make server side dimensions with custom settings different from overworld? (this things are possible for a very long time now with spigot for example and i do not see why this is not supported by forge)
I been hosting and playing forge modpacks for over 6 years now and i always needed to use some kind of mcpc+, cauldron or thermos to be able to manage a public server and i have decided to use 100% forge only to prove to others that forge is all that is needed, but i am facing with some parts that are very inflexible like creating dimensions and i want to help with this.

diesieben07 commented:
Your explanation seems long, but it doesn't convey anything. What are these "settings" you want to change? Give a detailed list.

goreacraft commented:
I will give a few examples:
1) create a mining dimension: 
  - no animals
  - no monsters
  - no villages
  - all day
(the purpose is to give players another dimension to use quarries or other mass mining tools so they will not make large wholes and destroy the overworld)
2) create a market dimension
  - no monsters
  - no villages
( this dimension will have chunks assigned to players where they can build their market ans trade each other items in one place and organized)
3) creative dimension:
  - players are in creative
  - all day
  (where players can build nice buildings and show off or test and learn different mods)
4) tutorial dimension
  - no monsters
  - all day
 ( different mods setups with messages on how to use more complex mods like Thaumcraft, AE, IC2 etc...)
5) events dimension
 - settings adjusted based on events types
 - different mingames where players can just take a break from normal gameplay

diesieben07 commented:
> no animals
no monsters
no villages
all day
(the purpose is to give players another dimension to use quarries or other mass mining tools so they will not make large wholes and destroy the overworld)

These all do not need a custom World class. There are events for animal and monster spawns, there is an event for village generation. And a custom WorldProvider can change the time of day.

> no monsters
no villages
( this dimension will have chunks assigned to players where they can build their market ans trade each other items in one place and organized)

Again, no need for a custom World class, see above.

> players are in creative
all day

Again, no need for a custom World class. Creative can be handled with change dimension event.

> settings adjusted based on events types
different mingames where players can just take a break from normal gameplay

Back to being vague...


All this suggests to me that you haven't actually spent any time trying to do this with existing hooks and mechanics, as such this is a case of tech support for the forum.

goreacraft commented:
Thank you for your time,

I will close this
